source/utils/security.py

```diff
@@ -177,19 +193,93 @@ def isObjectAboveLockScreen(obj: "NVDAObjects.NVDAObject") -> bool:
 	):
 		return True
 
+	import appModuleHandler
 	runningAppModules = appModuleHandler.runningTable.values()
 	lockAppModule = next(filter(_isLockAppAndAlive, runningAppModules), None)
 
 	if lockAppModule is None:
 		# lockAppModule not running/registered by NVDA yet
-		log.debugWarning(
+		log.debug(
 			"lockAppModule not detected when Windows is locked. "
-			"Cannot detect if object is in lock app, considering object as insecure. "
+			"Cannot detect if object is in lock app, considering object as safe. "
 		)
-	elif lockAppModule is not None and obj.processID == lockAppModule.processID:
 		return True
 
-	return False
+	from NVDAObjects.window import Window
+	if not isinstance(obj, Window):
+		log.debug(
+			"Cannot detect if object is in lock app, considering object as safe. "
+		)
+		# must be a window to get its HWNDVal
+		return True
+
+	return _isObjectAboveLockScreenCheckZOrder(obj.windowHandle, lockAppModule.processID)
+
+
+def _isObjectAboveLockScreenCheckZOrder(objWindowHandle: int, lockAppModuleProcessId: int) -> bool:
+	"""
+	This is a risky hack.
+	If the order is incorrectly detected,
+	the Windows UX may become inaccessible
+	or secure information may become accessible.
+
+	If these functions fail, where possible,
+	NVDA should make NVDA objects accessible.
+	"""
+
+	def _isWindowLockApp(hwnd: winUser.HWNDVal) -> bool:
+		windowProcessId, _threadId = winUser.getWindowThreadProcessID(hwnd)
+		return windowProcessId == lockAppModuleProcessId
+
+	def _isNVDAObjectWindow(hwnd: winUser.HWNDVal) -> bool:
+		return hwnd == objWindowHandle
+
+	lockAppZIndex = _getWindowZIndex(_isWindowLockApp)
+	objectZIndex = _getWindowZIndex(_isNVDAObjectWindow)
+	lockAppZIndexCheck = _getWindowZIndex(_isWindowLockApp)
+	objectZIndexCheck = _getWindowZIndex(_isNVDAObjectWindow)
+	if lockAppZIndex != lockAppZIndexCheck or objectZIndex != objectZIndexCheck:
+		log.debugWarning("Order of Windows has changed during execution")

```

Comment:
By using absolute values, you've moved the comparison into an untested function. Instead of `_getWindowZIndex`, why not compare the relative positions of two HWNDs I.E.

```python
def isFirstAboveSecond(first: HwndVal, second: HwndVal) -> bool:
	""" Returns true if first is above second.
	"'"
	assert first != second
    	desktopWindow = winUser.getDesktopWindow()
	nextWindow = winUser.getTopWindow(desktopWindow)
	currentIndex = 0
        firstIndex = secondIndex = None
	while nextWindow:
		if nextWindow == first:
			if firstIndex is not None:
				raise Exception("First window detected twice")
			else:
				firstIndex = currentIndex
		elif nextWindow == second:
			if secondIndex is not None:
				raise Exception("Second window detected twice")
			else:
				secondIndex = currentIndex
		if firstWindow is not None and secondWindow is not None:
			return firstIndex < secondIndex

		nextWindow = winUser.getWindow(nextWindow, winUser.GW_HWNDNEXT)
		index += 1
	if firstIndex is None and secondIndex is None:
		raise Exception("Both windows not detected")
	elif firstIndex is None:
		raise Exception("First window not detected")
	elif secondIndex is None:
		raise Exception("Second window not detected")
```

Reply:
I'd only run it a second time if an exception is raised.

Reply:
Testing this would be much more complex.
You'd have to consider windows moving between more states.
e.g. test when a window moves from `in front/behind` `first/second window` to `in front/behind` of `first/second window`, 
test when `first/second window` moves to `in front/behind` `first/second window` `before/after/when` reaching `first/second window`.

Reply:
But then the current approach doesn't test this complexity. What we care about is the relative z-order of two windows, and the potential for either window to move is inherent complexity.